File: AtCoder/codefes2016tr3_a.py
"""
・蟻本演習4-4-3
・貰うDP高速化、スライド最小値
・普通にやるとN^2*Kになって間に合わないのを、遷移のNを高速化するやつ。
　数え上げだと累積和で前計算とかできるけど、min,maxならセグ木使ったりになる。
・これはN*Kで既に計算量が10万*300=3000万あって、ここにセグ木のlogが乗ると、普通は通らない。
　で、そこを線形で処理するスライド最小値を使う。
・最初答え合わなくてあれー？ってなったけど、スライド最小値の実装が、真上からの遷移(何もしない遷移)
　も含むからで、今回はこれを含めたくないので、そのためにスライド最小値から戻ってきた時に
　位置を1つ右にずらしてる。(一番左は-INFで無効にする)
・多分pythonのdequeがそんなに速くないのとかが原因ぽいと思うんだけど、pypyでも通らないのでこの後C++で。
"""

import sys

def input(): return sys.stdin.readline().strip()
def list2d(a, b, c): return [[c] * b for i in range(a)]
def list3d(a, b, c, d): return [[[d] * c for j in range(b)] for i in range(a)]
def list4d(a, b, c, d, e): return [[[[e] * d for k in range(c)] for j in range(b)] for i in range(a)]
def ceil(x, y=1): return int(-(-x // y))
def INT(): return int(input())
def MAP(): return map(int, input().split())
def LIST(N=None): return list(MAP()) if N is None else [INT() for i in range(N)]
def Yes(): print('Yes')
def No(): print('No')
def YES(): print('YES')
def NO(): print('NO')
sys.setrecursionlimit(10**6)
INF = 10**18
MOD = 10**9 + 7
EPS = 10**-10

class SegTree:
    """
    セグメント木
    1.update: i番目の値をxに更新する
    2.query: 区間[l, r)の値を得る
    """

    def __init__(self, n, func, intv, A=[]):
        """
        :param n: 要素数(0-indexed)
        :param func: 値の操作に使う関数(min, max, add, gcdなど)
        :param intv: 要素の初期値(単位元)
        :param A: 初期化に使うリスト(オプション)
        """
        self.n = n
        self.func = func
        self.intv = intv
        # nより大きい2の冪数
        n2 = 1
        while n2 < n:
            n2 <<= 1
        self.n2 = n2
        self.tree = [self.intv] * (n2 << 1)
        # 初期化の値が決まっている場合
        if A:
            # 1段目(最下段)の初期化
            for i in range(n):
                self.tree[n2+i] = A[i]
            # 2段目以降の初期化
            for i in range(n2-1, -1, -1):
                self.tree[i] = self.func(self.tree[i*2], self.tree[i*2+1])

    def update(self, i, x):
        """
        i番目の値をxに更新
        :param i: index(0-indexed)
        :param x: update value
        """
        i += self.n2
        self.tree[i] = x
        while i > 0:
            i >>= 1
            self.tree[i] = self.func(self.tree[i*2], self.tree[i*2+1])

    def add(self, i, x):
        self.update(i, self.get(i) + x)

    def query(self, a, b):
        """
        [a, b)の値を得る
        :param a: index(0-indexed)
        :param b: index(0-indexed)
        """
        l = a + self.n2
        r = b + self.n2
        s = self.intv
        while l < r:
            if r & 1:
                r -= 1
                s = self.func(s, self.tree[r])
            if l & 1:
                s = self.func(s, self.tree[l])
                l += 1
            l >>= 1
            r >>= 1
        return s

    def get(self, i):
        """ 一点取得 """
        return self.tree[i+self.n2]

    def all(self):
        """ 全区間[0, n)の取得 """
        return self.tree[1]

    def print(self):
        for i in range(self.n):
            if i == self.n-1:
                print(self.get(i))
            else:
                print(self.get(i), end=' ')

# スライド最小値(比較関数、元の数列、遷移回数、遷移幅、遷移の重み)
def slide_min(func, A, k, w=1, v=0):
    from collections import deque

    N = len(A)
    res = A[:]
    for a in range(w):
        que = deque()
        i = 0
        while i*w+a < N:
            val = A[i*w+a] - i*v
            while que and func(que[-1][1], val) == val:
                que.pop()
            que.append((i, val))
            res[i*w+a] = que[0][1] + i*v
            if que[0][0] == i-k:
                que.popleft()
            i += 1
    return res

N, M, K = MAP()
A = LIST()

# dp = list2d(K+1, N, -INF)
# for i in range(N):
#     dp[1][i] = A[i]
# for i in range(2, K+1):
#     for j in range(N):
#         for k in range(max(j-M, 0), j):
#             dp[i][j] = max(dp[i][j], dp[i-1][k] + A[j]*i)
# ans = max(dp[K])
# print(ans)

# seg = SegTree(N, max, -INF, A)
# for i in range(2, K+1):
#     for j in range(N-1, -1, -1):
#         mx = seg.query(max(j-M, 0), j)
#         seg.update(j, mx + A[j]*i)
# ans = seg.all()
# print(ans)

dp = [0] * N
for i in range(N):
    dp[i] = A[i]
for i in range(2, K+1):
    dp = [-INF] + slide_min(max, dp[:-1], M-1, 1)
    for j in range(i-1, N):
        dp[j] += A[j]*i
ans = max(dp)
print(ans)

File: AtCoder/codefes2016tr3_a.cpp
/**
 * ・蟻本演習4-4-3
 * ・貰うDP高速化、スライド最小値
 * ・能書きはpythonの方にいっぱい書いたのでここには書かないけど、こっちは無事AC0.46秒。
 */

// #pragma GCC target("avx2")
// #pragma GCC optimize("O3")
// #pragma GCC optimize("unroll-loops")

#include <bits/stdc++.h>
using namespace std;

using ll = long long;
using ld = long double;
using pll = pair<ll, ll>;
using pii = pair<int, int>;
using vvl = vector<vector<ll>>;
using vvi = vector<vector<int>>;
using vvpll = vector<vector<pll>>;
#define rep(i, a, b) for (ll i=(a); i<(b); i++)
#define rrep(i, a, b) for (ll i=(a); i>(b); i--)
#define pb push_back
#define tostr to_string
#define ALL(A) A.begin(), A.end()
constexpr ll INF = LONG_LONG_MAX;
constexpr ll MOD = 1000000007;

template<typename T> vector<vector<T>> list2d(int N, int M, T init) { vector<vector<T>> res(N, vector<T>(M, init)); return res; }
template<typename T> vector<vector<vector<T>>> list3d(int N, int M, int L, T init) { vector<vector<vector<T>>> res(N, vector<vector<T>>(M, vector<T>(L, init))); return res; }

void print(ld out) { cout << fixed << setprecision(15) << out << '\n'; }
void print(double out) { cout << fixed << setprecision(15) << out << '\n'; }
template<typename T> void print(T out) { cout << out << '\n'; }
template<typename T1, typename T2> void print(pair<T1, T2> out) { cout << out.first << ' ' << out.second << '\n'; }
template<typename T> void print(vector<T> A) { rep(i, 0, A.size()) { cout << A[i]; cout << (i == A.size()-1 ? '\n' : ' '); } }
template<typename T> void print(set<T> S) { vector<T> A(S.begin(), S.end()); print(A); }

void Yes() { print("Yes"); }
void No() { print("No"); }
void YES() { print("YES"); }
void NO() { print("NO"); }

ll floor(ll a, ll b) { if (a < 0) { return (a-b+1) / b; } else { return a / b; } }
ll ceil(ll a, ll b) { if (a >= 0) { return (a+b-1) / b; } else { return a / b; } }
pll divmod(ll a, ll b) { ll d = a / b; ll m = a % b; return {d, m}; }
template<typename T> bool chmax(T &x, T y) { return (y > x) ? x = y, true : false; }
template<typename T> bool chmin(T &x, T y) { return (y < x) ? x = y, true : false; }

template<typename T> T sum(vector<T> A) { T res = 0; for (T a: A) res += a; return res; }
template<typename T> T max(vector<T> A) { return *max_element(ALL(A)); }
template<typename T> T min(vector<T> A) { return *min_element(ALL(A)); }

ll toint(string s) { ll res = 0; for (char c : s) { res *= 10; res += (c - '0'); } return res; }
int toint(char num) { return num - '0'; }
char tochar(int num) { return '0' + num; }
int ord(char c) { return (int)c; }
char chr(int a) { return (char)a; }

ll pow(int x, ll n) { ll res = 1; rep(_, 0, n) res *= x; return res; }
ll pow(ll x, ll n, int mod) { ll res = 1; while (n > 0) { if (n & 1) { res = (res * x) % mod; } x = (x * x) % mod; n >>= 1; } return res; }

int popcount(ll S) { return __builtin_popcountll(S); }
ll gcd(ll a, ll b) { return __gcd(a, b); }

// スライド最小値(比較関数、元の数列、遷移回数、遷移幅、遷移の重み)
template<typename T, typename F>
vector<T> slide_min(const F &func, vector<T> &A, ll k, ll w=1, ll v=0) {

    ll N = A.size();
    auto res = A;
    rep(a, 0, w) {
        deque<pair<int, T>> que;
        int i = 0;
        while (i*w+a < N) {
            ll val = A[i*w+a] - i*v;
            while (!que.empty() and func(que.back().second, val) == val) {
                que.pop_back();
            }
            que.pb({i, val});
            res[i*w+a] = que.front().second + i*v;
            if (que.front().first == i-k) {
                que.pop_front();
            }
            i++;
        }
    }
    return res;
}

int main() {
    cin.tie(0);
    ios::sync_with_stdio(false);

    ll N, M, K;
    cin >> N >> M >> K;
    vector<ll> A(N);
    rep(i, 0, N) cin >> A[i];

    vector<ll> dp(N);
    rep(i, 0, N) {
        dp[i] = A[i];
    }
    rep(i, 2, K+1) {
        dp.pop_back();
        dp = slide_min([](ll a, ll b) { return max(a, b); }, dp, M-1, 1);
        dp.insert(dp.begin(), -INF);
        rep(j, i-1, N) {
            dp[j] += A[j]*i;
        }
    }
    ll ans = max(dp);
    print(ans);
    return 0;
}
